internal/pkg/agent/application/actions/handlers/handler_action_upgrade.go

```diff
@@ -51,6 +54,25 @@ func (h *Upgrade) Handle(ctx context.Context, a fleetapi.Action, ack acker.Acker
 	if !runAsync {
 		return nil
 	}
+
+	if h.tamperProtectionFn() {
+		// Find inputs that want to receive UPGRADE action
+		// Endpoint needs to receive a signed UPGRADE action in order to be able to uncontain itself
+		state := h.coord.State()
+		ucs := findMatchingUnitsByActionType(state, a.Type())
+		if len(ucs) > 0 {
+			h.log.Debugf("handlerUpgrade: proxy/dispatch action '%+v'", a)
+			err := notifyUnitsOfProxiedAction(ctx, h.log, action, ucs, h.coord.PerformAction)
+			h.log.Debugf("handlerUpgrade: after action dispatched '%+v', err: %v", a, err)
+			if err != nil {
```

Comment:
Make sure that the error here is logged at the error level so we can always debug this as a source of upgrade failures.

CC @ycombinator, we will want to make sure we have a clear way to indicate that an upgrade failed because the action was not properly signed.